Java/core/src/main/java/com/amazon/randomcutforest/preprocessor/ImputePreprocessor.java

```diff
@@ -381,54 +242,59 @@ protected int determineGap(long timestampGap, double averageGap) {
         }
     }
 
+    public int numberOfImputes(long timestamp) {
+        long lastInputTimeStamp = previousTimeStamps[shingleSize - 1];
+        return determineGap(timestamp - lastInputTimeStamp, getTimeFactor(timeStampDeviations[1])) - 1;
+    }
+
     /**
      * a single function that constructs the next shingle, with the option of
      * committing them to the forest However the shingle needs to be generated
      * before we process a point; and can only be committed once the point has been
-     * scored. Having the same deterministic transformation is essential
+     * scored. Having the same deterministic transformation can be useful. Note for
+     * this imputation timestamp cannot be missing
      *
-     * @param descriptor   description of the current point
      * @param averageGap   the gap in timestamps
      * @param changeForest boolean determining if we commit to the forest or not
      * @param forest       the resident RCF
      * @return the next shingle
      */
-    protected double[] generateShingle(AnomalyDescriptor descriptor, double averageGap, boolean changeForest,
-            RandomCutForest forest) {
-        double[] input = descriptor.getCurrentInput();
-        long timestamp = descriptor.getInputTimestamp();
+    protected float[] generateShingle(double[] inputTuple, long timestamp, int[] missingValues, double averageGap,
+            boolean changeForest, RandomCutForest forest) {
         long lastInputTimeStamp = previousTimeStamps[shingleSize - 1];
-        int[] missingValues = descriptor.getMissingValues();
-        checkArgument(missingValues == null || (imputationMethod != LINEAR && imputationMethod != NEXT),
-                " cannot perform imputation on most recent missing value with this method");
-        /*
-         * Note only ZERO, FIXED_VALUES, PREVIOUS and RCF are reasonable options if
-         * missing values are present.
-         */
-
-        checkArgument(internalTimeStamp > 0, "imputation should have forced normalization");
-        double[] savedInput = getShingledInput(shingleSize - 1);
-
-        // previous value should be defined
-        double[] previous = new double[inputLength];
-        System.arraycopy(lastShingledInput, lastShingledInput.length - inputLength, previous, 0, inputLength);
+        double[] input = Arrays.copyOf(inputTuple, inputLength);
+        double[] previous = getShingledInput(shingleSize - 1);
+
+        if (missingValues != null && imputationMethod != RCF) {
+            // for STREAMING_IMPUTE the timestamp cannot be missing
+            // hence missingValues[] can be 0 to inputLength - 1
+            // for next and Linear there are no current values
```

Comment:
For linear, can we use previous n value at dimension i to do linear interpolation?

When will next imputation be useful? I can only think of historical AD as next value is available.

Current approach defaults to fixedvalues or previous seems to silently make a decision for the customer. Should we fail fast and return empty shingle since current imputation strategy does not work?

Reply:
I saw you imputed 3 times:

First, impute current input value:
```
if (missingValues != null && imputationMethod != RCF) {
            // for STREAMING_IMPUTE the timestamp cannot be missing
            // hence missingValues[] can be 0 to inputLength - 1
            // for next and Linear there are no current values
            // we are forced to use fixedvalues or previous
            for (int missingValue : missingValues) {
                input[missingValue] = (defaultFill == null) ? previous[missingValue] : defaultFill[missingValue];
            }
        }
```

Then impute the missing gaps between last seen value and the value immediately before current input:

```
double[] savedInput = Arrays.copyOf(previous, inputLength);
        // using the global dependency
        int numberToImpute = determineGap(timestamp - lastInputTimeStamp, averageGap) - 1;
        if (numberToImpute > 0) {
            double step = 1.0 / (numberToImpute + 1);
            // the last impute corresponds to the current observed value
            for (int i = 0; i < numberToImpute; i++) {
                // only the last tuple is partial
                double[] result = impute(input, missingValues, false, step * (i + 1), previous, forest);
                updateForest(changeForest, result, timestamp, forest, true);
            }
        }
```

Then you impute current input value again:

```
 boolean ignoreImpute = (missingValues == null) || (missingValues.length == 0);

        double[] newInput = (ignoreImpute) ? input : impute(input, missingValues, true, 0, previous, forest);
 ```
 
 Is my understanding correct? If yes, why do we need to impute current input again?

Reply:
Good q. The first part  corresponds to missingValues != null && imputationMethod != RCF.

This means that we have a partial current tuple. We cannot define next/linear for partial current tuples -- even though the next/linear is applied to the intervening tuples (hence minus 1). These tuples are imputed fully (not partial). The first step above is very cheap operation. The second block can use any method (including RCF).

Then finally we come back to current tuple. If there were indeed missing values (ignoreImpute = false) then we need to impute again (specially for RCF) -- and this impute is not (necessarily) cheap. Observe that previous could have changed due to the intervening block

The reason for separating into three parts was that we could have complete control and restore the changes if we ever wanted to. Because anomalyScore() can be posed on a partial tuple; and then update occurs with a different tuple.

All that being said, this does look clunky. Let me see if there ia an easier way. But we do have to separate the invocations where we make up a fully missing record and where we know some values (current tuple).

Reply:
Changed -- added the fast forward. But I think we need to move more changes to any subsequent PRs ... this seems to be much better than the 3.8 version of the preprocessor already.